### PR TITLE
Fix arena access violation for -profile option

### PR DESCRIPTION
**Pull request by ThomsonTan**
Fixed -profile for NativeCodeGenerator, disabled it for GC phase as workaround.

### CODE REVIEW

File: lib/Common/Memory/Recycler.cpp [hunk 1]
```diff
@@ -5240,6 +5240,7 @@ Recycler::FinishConcurrentCollect(CollectionFlags flags)
 
 #ifdef PROFILE_EXEC
     Js::Phase concurrentPhase = Js::ConcurrentCollectPhase;
+    static_cast<Js::Phase>(concurrentPhase);
```
Comment (curtisman) [inline lib/Common/Memory/Recycler.cpp]: What is this line for?

File: lib/Common/Memory/Recycler.h [hunk 2]
```diff
@@ -2372,7 +2375,8 @@ struct ForceLeafAllocator<RecyclerNonLeafAllocator>
     typedef RecyclerLeafAllocator AllocatorType;
 };
 
-#ifdef PROFILE_EXEC
+// TODO: enable -profile for GC phases
+#if defined(PROFILE_EXEC) && 0
```
Comment (curtisman) [inline lib/Common/Memory/Recycler.h]: What is blocking from making this work? 
AFAIK, this used to work, but something might have broke it?

File: lib/Common/Memory/HeapInfo.cpp [hunk 3]
```diff
@@ -881,6 +881,7 @@ HeapInfo::AddLargeHeapBlock(size_t size)
 void HeapInfo::SweepBuckets(RecyclerSweep& recyclerSweep, bool concurrent)
 {
     Recycler * recycler = recyclerSweep.GetRecycler();
+    static_cast<Recycler*>(recycler);
```
Comment (curtisman) [inline lib/Common/Memory/HeapInfo.cpp]: What is this line for?

File: lib/Common/Memory/Recycler.h [hunk 4]
```diff
@@ -979,6 +978,10 @@ class Recycler
     PageAllocator backgroundProfilerPageAllocator;
     DListBase<ArenaAllocator> backgroundProfilerArena;
 #endif
+
+    // destruct autoHeap after backgroundProfilerPageAllocator;
+    HeapInfo autoHeap;
```
Comment (curtisman) [inline lib/Common/Memory/Recycler.h]: Can you explain why this needs to be after backgroundProfilerPageAllocator.  Those seems to me to be unrelated?

File: lib/Backend/NativeCodeGenerator.cpp [hunk 5]
```diff
@@ -2802,25 +2810,40 @@ void
...
             codegenProfiler = codegenProfiler->next;
         }
     }
     else
     {
         //Merge all the codegenProfiler for single snapshot.
-        codegenProfiler = codegenProfiler->next;
-        while (codegenProfiler)
+        Js::ScriptContextProfiler* mergeToProfiler = codegenProfiler;
+        while (mergeToProfiler != nullptr && !mergeToProfiler->IsInitialized())
```
Comment (curtisman) [inline lib/Backend/NativeCodeGenerator.cpp]: Why an extra loop here to find the first initialized profiler?
The loop below should be able to handle that already.  
This loop seems redundant.

File: lib/Common/Memory/Recycler.cpp (hunk 1 above)
Comment (ThomsonTan) [inline lib/Common/Memory/Recycler.cpp]: As -profile is disabled for GC phases, this is workaround for the unreferenced local variable. I have plan to enable -profile for GC phases so this is temporary.

---
In reply to: [66280866](https://github.com/Microsoft/ChakraCore/pull/1096#discussion_r66280866) [](ancestors = 66280866)

File: lib/Common/Memory/Recycler.h (hunk 2 above)
Comment (ThomsonTan) [inline lib/Common/Memory/Recycler.h]: Background GC threads use the one profiler from the same recycler object. Profiler object is not thread safe (phase push/pop cannot match if accessed from multiple threads).

---
In reply to: [66281159](https://github.com/Microsoft/ChakraCore/pull/1096#discussion_r66281159) [](ancestors = 66281159)

**Comment by curtisman** [comment]
I think this change should go to master instead of release/1.2?

File: lib/Common/Memory/HeapInfo.cpp (hunk 3 above)
Comment (ThomsonTan) [inline lib/Common/Memory/HeapInfo.cpp]: The same as above, a temporary workaround for unreferenced local variable.

---
In reply to: [66281536](https://github.com/Microsoft/ChakraCore/pull/1096#discussion_r66281536) [](ancestors = 66281536)

File: lib/Backend/NativeCodeGenerator.cpp [hunk 6]
```diff
@@ -2802,25 +2810,40 @@ void
...
 {
     Js::ScriptContextProfiler *codegenProfiler = this->backgroundCodeGenProfiler;
+
     if (Js::Configuration::Global.flags.Verbose)
     {
         //Print individual CodegenProfiler information in verbose mode
         while (codegenProfiler)
         {
-            codegenProfiler->ProfilePrint(Js::Configuration::Global.flags.Profile.GetFirstPhase());
+            if (codegenProfiler->IsInitialized())
```
Comment (curtisman) [inline lib/Backend/NativeCodeGenerator.cpp]: This is already checked in ScriptContextProfiler::ProfilePrint.  No need to add a check here again

File: lib/Common/Memory/Recycler.h (hunk 4 above)
Comment (ThomsonTan) [inline lib/Common/Memory/Recycler.h]: backgroundProfilerArena allocated heap block from autoHeap. If autoHeap destructed before backgroundProfilerArena, the destructor for the latter will run on some freed heap blocks.

---
In reply to: [66281817](https://github.com/Microsoft/ChakraCore/pull/1096#discussion_r66281817) [](ancestors = 66281817)

File: lib/Backend/NativeCodeGenerator.cpp (hunk 5 above)
Comment (ThomsonTan) [inline lib/Backend/NativeCodeGenerator.cpp]: Below loop doesn't ensure the first profiler (mergeToProfiler) is initialized.

---
In reply to: [66282264](https://github.com/Microsoft/ChakraCore/pull/1096#discussion_r66282264) [](ancestors = 66282264)

**Comment by ThomsonTan** [comment]
Thanks for the reminder. I am experimenting JIT throughput based on release/1.2 so I try to push to this branch. Is there any other reason this should go to master?

---
In reply to: [224635925](https://github.com/Microsoft/ChakraCore/pull/1096#issuecomment-224635925) [](ancestors = 224635925)

File: lib/Backend/NativeCodeGenerator.cpp (hunk 6 above)
Comment (ThomsonTan) [inline lib/Backend/NativeCodeGenerator.cpp]: You are right, this check is redundant. No merge action here so the check in ScriptContextProfiler::ProfilePrint is enough.

---
In reply to: [66284121](https://github.com/Microsoft/ChakraCore/pull/1096#discussion_r66284121) [](ancestors = 66284121)

File: lib/Common/Memory/Recycler.h (hunk 4 above)
Comment (curtisman) [inline lib/Common/Memory/Recycler.h]: Arena doesn't allocate stuff in recycler's autoHeap.

**Pull request closed without merging**